x/incentive/keeper/keeper.go

```diff
@@ -39,10 +45,267 @@ func NewKeeper(
+}
+
+// Calculate the delegated amount
+func (k Keeper) CalculateDelegatedAmount(ctx sdk.Context, delegator string) sdk.Int {
+	// Derivate bech32 based delegator address
+	delAdr, err := sdk.AccAddressFromBech32(delegator)
+	if err != nil {
+		return sdk.ZeroInt()
+	}
+
+	// Get elys delegation for creator address
+	delegatedAmt := sdk.ZeroInt()
+
+	// Iterate all delegations for the specified delegator
+	k.stk.IterateDelegations(
+		ctx, delAdr,
+		func(_ int64, del stakingtypes.DelegationI) (stop bool) {
+			// Get validator address
+			valAddr := del.GetValidatorAddr()
+			// Get validator
+			val := k.stk.Validator(ctx, valAddr)
+
+			/****************************************************************************************/
+			// --------------------(Cosmos SDK staking module implementation)------------------------
+			// del.Shares = val.GetDelegatorShares().MulInt(amt).QuoInt(val.GetTokens())
+			// amt = del.Shares * val.GetTokens / val.GetDelegatorShares()
+			/****************************************************************************************/
+			amt := del.GetShares().MulInt(val.GetTokens()).Quo(val.GetDelegatorShares()).TruncateInt()
+
+			// Sum the individual delegation
+			delegatedAmt.Add(amt)
+			return false
+		},
+	)
+
+	return delegatedAmt
+}
+
+// Find out active incentive params
+func (k Keeper) FindProperIncentiveParm(ctx sdk.Context, epochIdentifier string) (bool, types.IncentiveInfo, types.IncentiveInfo) {
+	// Incentive params initialize
+	stakeIncentive := types.IncentiveInfo{}
+	lpIncentive := types.IncentiveInfo{}
+
+	// Current block timestamp
+	timestamp := ctx.BlockTime().Unix()
+	foundIncentive := false
+
+	// Fetch incentive params
+	params := k.GetParams(ctx)
+
+	// Find approporiate Incentive Info
+	// Consider epochIdentifier and start time
+	// Consider epochNumber as well
+	for _, ii := range params.StakeIncentives {
+		if ii.EpochIdentifier == epochIdentifier && timestamp >= ii.StartTime.Unix() {
+			stakeIncentive = ii
+			foundIncentive = true
+			break
+		}
+	}
+
+	// Find approporiate Incentive Info
+	for _, ii := range params.LpIncentives {
+		if ii.EpochIdentifier == epochIdentifier && timestamp >= ii.StartTime.Unix() {
+			lpIncentive = ii
+		}
+	}
+
+	// return found, stake, lp incentive params
+	return foundIncentive, stakeIncentive, lpIncentive
+}
+
+// Update uncommitted token amount
+// Called back through epoch hook
+func (k Keeper) UpdateUncommittedTokens(ctx sdk.Context, epochIdentifier string) {
+	// Recalculate total committed info
+	k.UpdateTotalCommitmentInfo(ctx, epochIdentifier)
+
+	// Iterate all delegations for the specified delegator
+	// Process to increase uncomitted token amount of Eden & Eden boost
+	k.cmk.IterateCommitments(
+		ctx, epochIdentifier,
+		func(commitments ctypes.Commitments) bool {
+			// Find out active incentive params
+			foundIncentive, stakeIncentive, lpIncentive := k.FindProperIncentiveParm(ctx, epochIdentifier)
```

Comment:
@kenta-elys We can pass these Incentive objects in through the hook so that they don't have to be found twice.

Reply:
Yes I could see this is inefficient. But If I should have to pass it as parameters, commitment module should have reference to incentive module while incentive module also refers it. And this causes an error in golang.

Reply:
Fixed.